sql/catalyst/src/main/scala/org/apache/spark/sql/util/PartitioningUtils.scala

```diff
@@ -30,14 +33,20 @@ object PartitioningUtils {
    */
   def normalizePartitionSpec[T](
       partitionSpec: Map[String, T],
-      partColNames: Seq[String],
+      partCols: StructType,
       tblName: String,
       resolver: Resolver): Map[String, T] = {
     val normalizedPartSpec = partitionSpec.toSeq.map { case (key, value) =>
-      val normalizedKey = partColNames.find(resolver(_, key)).getOrElse {
+      val normalizedFiled = partCols.find(f => resolver(f.name, key)).getOrElse {
         throw new AnalysisException(s"$key is not a valid partition column in table $tblName.")
       }
-      normalizedKey -> value
+
+      val normalizedVal = normalizedFiled.dataType match {
+        case CharType(len) if value != null =>
+          StringUtils.rightPad(value.toString, len).asInstanceOf[T]
```

Comment:
T can be `Option[String]`, and calling `toString` blindly seems risky.

Reply:
we probably should do a pattern match and handle `String` `Option[String]` differently.

Reply:
shall we call `charTypeWriteSideCheck` here?

Reply:
Here for the session catalog, we will have null and `__HIVE_DEFAULT_PARTITION__` here, what's the behavior for v2 tables?

Reply:
`__HIVE_DEFAULT_PARTITION__ ` is the same as `null`, and we pass them to v2 catalog as it is.